File: c++/src/dynamixel_sdk/dynamixel_api/motor.cpp
// Copyright 2025 ROBOTIS CO., LTD.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Hyungyu Kim

#include "dynamixel_api/motor.hpp"
#include "dynamixel_api/connector.hpp"

namespace dynamixel
{
Motor::Motor(uint8_t id, uint16_t model_number, Connector * connector)
: id_(id),
  model_number_(model_number),
  model_name_(ControlTable::getModelName(model_number)),
  connector_(connector),
  control_table_(ControlTable::getControlTable(model_number))
{}

Motor::~Motor() {}

Result<void, DxlError> Motor::enableTorque()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Torque Enable");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 1);
  return result;
}

Result<void, DxlError> Motor::disableTorque()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Torque Enable");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 0);
  return result;
}

Result<void, DxlError> Motor::setGoalPosition(uint32_t position)
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Goal Position");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  if (item.size != 4) {
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  }
  Result<void, DxlError> result = connector_->write4ByteData(id_, item.address, position);
  return result;
}

Result<void, DxlError> Motor::setGoalVelocity(uint32_t velocity)
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Goal Velocity");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  if (item.size != 4) {
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  }
  Result<void, DxlError> result = connector_->write4ByteData(id_, item.address, velocity);
  return result;
}

Result<void, DxlError> Motor::LEDOn()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("LED");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 1);
  return result;
}

Result<void, DxlError> Motor::LEDOff()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("LED");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 0);
  return result;
}

Result<uint16_t, DxlError> Motor::ping()
{
  Result<uint16_t, DxlError> result = connector_->read2ByteData(id_, 0);
  return result;
}

Result<uint8_t, DxlError> Motor::isTorqueOn()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Torque Enable");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<uint8_t, DxlError> result = connector_->read1ByteData(id_, item.address);
  return result;
}

Result<uint8_t, DxlError> Motor::isLEDOn()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("LED");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<uint8_t, DxlError> result = connector_->read1ByteData(id_, item.address);
  return result;
}

Result<int32_t, DxlError> Motor::getPresentPosition()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Present Position");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<uint32_t, DxlError> result = connector_->read4ByteData(id_, item.address);
  if (!result.isSuccess()) {
    return result.error();
  }
  return static_cast<int32_t>(result.value());
}

Result<int32_t, DxlError> Motor::getPresentVelocity()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Present Velocity");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<uint32_t, DxlError> result = connector_->read4ByteData(id_, item.address);
  if (!result.isSuccess()) {
    return result.error();
  }
  return static_cast<int32_t>(result.value());
}

Result<uint32_t, DxlError> Motor::getMaxPositionLimit()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Max Position Limit");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  if (item.size != 4) {
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  }
  Result<uint32_t, DxlError> result = connector_->read4ByteData(id_, item.address);
  return result;
}

Result<uint32_t, DxlError> Motor::getMinPositionLimit()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Min Position Limit");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  if (item.size != 4) {
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  }
  Result<uint32_t, DxlError> result = connector_->read4ByteData(id_, item.address);
  return result;
}

Result<uint32_t, DxlError> Motor::getVelocityLimit()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Velocity Limit");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  if (item.size != 4) {
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  }
  Result<uint32_t, DxlError> result = connector_->read4ByteData(id_, item.address);
  return result;
}

Result<void, DxlError> Motor::changeID(uint8_t new_id)
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("ID");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, new_id);
  if (result.isSuccess()) {
    id_ = new_id;
  }
  return result;
}

Result<void, DxlError> Motor::changeBaudRate(int new_baud_rate)
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Baud Rate");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  uint8_t baud_value = static_cast<uint8_t>(2000000 / new_baud_rate - 1);
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, baud_value);
  return result;
}

Result<void, DxlError> Motor::setPositionControlMode()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Operating Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 3);
  return result;
}

Result<void, DxlError> Motor::setVelocityControlMode()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Operating Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 1);
  return result;
}

Result<void, DxlError> Motor::setCurrentControlMode()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Operating Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, 0);
  return result;
}

Result<void, DxlError> Motor::setTimeBasedProfile()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Drive Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  auto read_result = connector_->read1ByteData(id_, item.address);
  if (!read_result.isSuccess()) {
    return read_result.error();
  }
  uint8_t drive_mode = read_result.value();
  drive_mode |= 0b00000100;
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, drive_mode);
  return result;
}

Result<void, DxlError> Motor::setVelocityBasedProfile()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Drive Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  auto read_result = connector_->read1ByteData(id_, item.address);
  if (!read_result.isSuccess()) {
    return read_result.error();
  }
  uint8_t drive_mode = read_result.value();
  drive_mode &= 0b11111011;
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, drive_mode);
  return result;
}

Result<void, DxlError> Motor::setNormalDirection()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Drive Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  auto read_result = connector_->read1ByteData(id_, item.address);
  if (!read_result.isSuccess()) {
    return read_result.error();
  }
  uint8_t drive_mode = read_result.value();
  drive_mode &= 0b11111110;
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, drive_mode);
  return result;
}

Result<void, DxlError> Motor::setReverseDirection()
{
  Result<ControlTableItem, DxlError> item_result = getControlTableItem("Drive Mode");
  if (!item_result.isSuccess()) {
    return item_result.error();
  }
  const ControlTableItem & item = item_result.value();
  auto read_result = connector_->read1ByteData(id_, item.address);
  if (!read_result.isSuccess()) {
    return read_result.error();
  }
  uint8_t drive_mode = read_result.value();
  drive_mode |= 0b00000001;
  Result<void, DxlError> result = connector_->write1ByteData(id_, item.address, drive_mode);
  return result;
}

Result<void, DxlError> Motor::reboot()
{
  Result<void, DxlError> result = connector_->reboot(id_);
  return result;
}

Result<void, DxlError> Motor::factoryResetAll()
{
  Result<void, DxlError> result = connector_->factoryReset(id_, 0xFF);
  return result;
}

Result<void, DxlError> Motor::factoryResetExceptID()
{
  Result<void, DxlError> result = connector_->factoryReset(id_, 0x01);
  return result;
}

Result<void, DxlError> Motor::factoryResetExceptIDAndBaudRate()
{
  Result<void, DxlError> result = connector_->factoryReset(id_, 0x02);
  return result;
}

Result<ControlTableItem, DxlError> Motor::getControlTableItem(const std::string & name)
{
  auto it = control_table_.find(name);
  if (it == control_table_.end())
    return DxlError::API_FUNCTION_NOT_SUPPORTED;
  return it->second;
}
}  // namespace dynamixel

File: python/src/dynamixel_easy_sdk/motor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

################################################################################
# Copyright 2025 ROBOTIS CO., LTD.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

# Author: Hyungyu Kim

from dynamixel_easy_sdk.dynamixel_error import *
from dynamixel_easy_sdk.control_table import ControlTable
from enum import IntEnum

class OperatingMode(IntEnum):
    POSITION = 3
    VELOCITY = 1
    CURRENT = 0
    PWM = 16
    EXTENDED_POSITION = 4

class Direction(IntEnum):
    NORMAL = 0
    REVERSE = 1

class ProfileConfiguration(IntEnum):
    VELOCITY_BASED = 0
    TIME_BASED = 1

class Motor:
    def __init__(self, motor_id: int, model_number: int, connector):
        self.id = motor_id
        self.model_number = model_number
        self.model_name = ControlTable.getModelName(model_number)
        self.connector = connector
        self.control_table = ControlTable.getControlTable(model_number)
        self.torque_status = self.isTorqueOn()
        self.operating_mode_status = self.getOperatingMode()

    def enableTorque(self) -> None:
        item = self._getControlTableItem("Torque Enable")
        self.connector.writeData(self.id, item.address, item.size, 1)
        self.torque_status = 1

    def disableTorque(self) -> None:
        item = self._getControlTableItem("Torque Enable")
        self.connector.writeData(self.id, item.address, item.size, 0)
        self.torque_status = 0

    def setGoalPosition(self, position: int) -> None:
        self._checkTorqueStatus(1)
        if self.operating_mode_status != OperatingMode.POSITION and self.operating_mode_status != OperatingMode.EXTENDED_POSITION:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_OPERATING_MODE_MISMATCH)
        item = self._getControlTableItem("Goal Position")
        self.connector.writeData(self.id, item.address, item.size, position)

    def setGoalVelocity(self, velocity: int) -> None:
        self._checkTorqueStatus(1)
        self._checkOperatingModeStatus(OperatingMode.VELOCITY)
        item = self._getControlTableItem("Goal Velocity")
        self.connector.writeData(self.id, item.address, item.size, velocity)

    def setGoalCurrent(self, current: int) -> None:
        self._checkTorqueStatus(1)
        self._checkOperatingModeStatus(OperatingMode.CURRENT)
        item = self._getControlTableItem("Goal Current")
        self.connector.writeData(self.id, item.address, item.size, current)

    def setGoalPwm(self, pwm: int) -> None:
        self._checkTorqueStatus(1)
        self._checkOperatingModeStatus(OperatingMode.PWM)
        item = self._getControlTableItem("Goal PWM")
        self.connector.writeData(self.id, item.address, item.size, pwm)

    def ledOn(self) -> None:
        item = self._getControlTableItem("LED")
        self.connector.writeData(self.id, item.address, item.size, 1)

    def ledOff(self) -> None:
        item = self._getControlTableItem("LED")
        self.connector.writeData(self.id, item.address, item.size, 0)

    def ping(self) -> int:
        value = self.connector.ping(self.id)
        return value

    def isTorqueOn(self) -> int:
        item = self._getControlTableItem("Torque Enable")
        self.torque_status = self.connector.readData(self.id, item.address, item.size)
        return self.torque_status

    def isLedOn(self) -> int:
        item = self._getControlTableItem("LED")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getPresentPosition(self) -> int:
        item = self._getControlTableItem("Present Position")
        unsigned_value = self.connector.readData(self.id, item.address, item.size)
        return self._toSignedInt(unsigned_value, item.size)

    def getPresentVelocity(self) -> int:
        item = self._getControlTableItem("Present Velocity")
        unsigned_value = self.connector.readData(self.id, item.address, item.size)
        return self._toSignedInt(unsigned_value, item.size)

    def getMaxPositionLimit(self) -> int:
        item = self._getControlTableItem("Max Position Limit")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getMinPositionLimit(self) -> int:
        item = self._getControlTableItem("Min Position Limit")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getVelocityLimit(self) -> int:
        item = self._getControlTableItem("Velocity Limit")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getCurrentLimit(self) -> int:
        item = self._getControlTableItem("Current Limit")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getPwmLimit(self) -> int:
        item = self._getControlTableItem("PWM Limit")
        value = self.connector.readData(self.id, item.address, item.size)
        return value

    def getOperatingMode(self) -> OperatingMode:
        item = self._getControlTableItem("Operating Mode")
        value = self.connector.readData(self.id, item.address, item.size)
        self.operating_mode_status = OperatingMode(value)
        return self.operating_mode_status

    def changeId(self, newId: int) -> None:
        self._checkTorqueStatus(0)
        item = self._getControlTableItem("ID")
        self.connector.writeData(self.id, item.address, item.size, newId)
        self.id = newId

    def setOperatingMode(self, mode: OperatingMode) -> None:
        self._checkTorqueStatus(0)
        item = self._getControlTableItem("Operating Mode")
        self.connector.writeData(self.id, item.address, item.size, int(mode))
        self.operating_mode_status = mode

    def setProfileConfiguration(self, config: ProfileConfiguration) -> None:
        self._checkTorqueStatus(0)
        item = self._getControlTableItem("Drive Mode")
        mode_value = self.connector.readData(self.id, item.address, item.size)

        PROFILE_BIT = 0b00000100
        if config == ProfileConfiguration.TIME_BASED:
            mode_value |= PROFILE_BIT
        else:
            mode_value &= ~PROFILE_BIT

        self.connector.writeData(self.id, item.address, item.size, mode_value)

    def setDirection(self, direction: Direction) -> None:
        self._checkTorqueStatus(0)
        item = self._getControlTableItem("Drive Mode")
        mode_value = self.connector.readData(self.id, item.address, item.size)

        DIR_BIT = 0b00000001
        if direction == Direction.NORMAL:
            mode_value &= ~DIR_BIT
        else:
            mode_value |= DIR_BIT

        self.connector.writeData(self.id, item.address, item.size, mode_value)

    def setPositionPGain(self, value: int) -> None: self._writeGain("Position P Gain", value)
    def setPositionIGain(self, value: int) -> None: self._writeGain("Position I Gain", value)
    def setPositionDGain(self, value: int) -> None: self._writeGain("Position D Gain", value)
    def setVelocityPGain(self, value: int) -> None: self._writeGain("Velocity P Gain", value)
    def setVelocityIGain(self, value: int) -> None: self._writeGain("Velocity I Gain", value)

    def _getControlTableItem(self, name):
        item = self.control_table.get(name)
        if item is None:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_FUNCTION_NOT_SUPPORTED)
        return item

    def _checkTorqueStatus(self, status: int):
        if self.torque_status != status:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_TORQUE_STATUS_MISMATCH)

    def _checkOperatingModeStatus(self, mode: OperatingMode):
        if self.operating_mode_status != mode:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_OPERATING_MODE_MISMATCH)

    def _writeGain(self, name, value):
        self._checkTorqueStatus(0)
        item = self._getControlTableItem(name)
        self.connector.writeData(self.id, item.address, item.size, value)

    def _toSignedInt(self, value: int, size: int) -> int:
        bits = size * 8
        if value >= (1 << (bits - 1)):
            value -= (1 << bits)
        return value
